### ISSUE TITLE
Suppressing Rethrow

### ISSUE THREAD

**Issue by dgauerke** [vescon/MethodBoundaryAspect.Fody#2]
Under certain circumstances it would be nice to not rethrow the exception.  Without changing the code too much, just adding an if statement like so would be great:

```
        object[] objArray = new object[0];
        MethodExecutionArgs args = new MethodExecutionArgs();
        args.Arguments = objArray;
        MethodBase currentMethod = MethodBase.GetCurrentMethod();
        args.Method = currentMethod;
        ExceptionAspect exceptionAspect = new ExceptionAspect();
        exceptionAspect.OnEntry(args);
        try
        {
            throw (Exception)new NullReferenceException("Test");
        }
        catch (Exception ex)
        {
            args.Exception = ex;
            exceptionAspect.OnException(args);
            if (args != null)
                throw;
        }
```

**Comment by dgauerke** [comment]
Sorry, that should be
if (args.Exception != null)
    throw;

**Comment by marcells** [comment]
It that the way you intend to use the feature?

Generation:
```
...
catch (Exception ex)
{
    args.Exception = ex;
    exceptionAspect.OnException(args);
    if (args.FlowBehavior == FlowBehavior.RethrowException)
        throw;
}
```

```
enum FlowBehavior
{
  Default,
  Continue,
  RethrowException
}
```

The aspect:
```
class MyAspect : OnMethodBoundaryAspect
{
  public override void OnException(MethodExecutionArgs args)
  {
    if(args.Exception.GetType() == typeof(MyException))
      args.FlowBehavior = FlowBehavior.Continue;

    // default FlowBehavior will stay FlowBehavior.RethrowException;
  }
}
```

**Comment by dgauerke** [comment]
Yes, that sounds like what I’d like.  I just needed some way of not re-throwing in the aspect.

Thanks,
David Gauerke

**Comment by sandimschuh** [comment]
Would be great to have this FlowBehavior. That would me possibly switch from PostSharp back to Fody :)

**Comment by marcells** [comment]
@ralf1108 is currently changing the internal implementation how the aspect is weaved. When he's finished, we will take a look at this issue.

**Comment by marcells** [comment]
Currently were are working on debugging issues of the aspect. Do you mind making a PR for this feature?

**Comment by clawrenceks** [comment]
I echo that thoughts of @sandimschuh here. I think FlowBehaviour is the only thing which I need to move from PostSharp over to this Aspect.

**Comment by marcells** [comment]
As a workaround, is it possible to use something like this in the `OnException(args)` method?

```
public override void OnException(MethodExecutionArgs args)
  {
    if(args.Exception.GetType() == typeof(MyException))
      return;

    throw args.Exception;
  }
```

It's not exactly the same stacktrace, but the result should be the same.

**Comment by clawrenceks** [comment]
Actually I think perhaps I am looking for something different.

All of the code in my scenario is async and what I was hoping to achieve was in the event of an exception, to swallow the exception and return a failed result object.

My failed result object would contain information I want a user to see in a UI. When I say swallow the exception, what I am really looking to do is prevent it from being throw (I want to log it using our exception logging solution).

Is there any way to achieve the above, given the async nature of what I am doing? In PostSharp 6 I would use an interceptor to achieve this type of behaviour.

**Comment by clawrenceks** [comment]
In case it adds additional context, the following shows how I achieve what I described above using an Interceptor in PostSharp 6.

I realise interception is different than decoration, so perhaps what I am doing below cannot be achieved using MethodBoundaryAspect?

```
public override async Task OnInvokeAsync(MethodInterceptionArgs args)
{
    try
    {
        await args.ProceedAsync();
    }
    catch (Exception)
    {
        //Log exception with exception logging solution
        //Set method return type to a result object with generic error to show in UI
    }
}
```

**Comment by marcells** [comment]
I'm sorry to say, that that's currently not implemented, that after OnExit/OnException the return value is not taken from the MethodExecutionArgs. You could try to implement it and make a PR.

Take a look [here](https://github.com/vescon/MethodBoundaryAspect.Fody/blob/cf0805630714e7c3d56064e8165c40154f447e97/MethodBoundaryAspect.Fody/MethodWeaver.cs#L67).

**Comment by keith-anders** [comment]
I just submitted a PR to change return values in OnExit. That should allow @clawrenceks 's use case, which would take care of the async case. I can work on the synchronous case this weekend or some time next week.

**Comment by Ralf1108** [comment]
this issue should be fixed by PR #43 
@dgauerke can you verify that your issue is fixed?
